formal spec overview for the 3rd exception handling proposal (#143)
WebAssembly/exception-handling: proposals/exception-handling/Exceptions-formal-overview.md

# 3rd proposal formal spec overview

This is an overview of the 3rd proposal's formal spec additions, to aid in discussions concerning the proposed semantics.

## Abstract Syntax

### Types

Exception Types

```
exntype ::= [t*] -> []
```

### Instructions

```
instr ::= ... | throw x | rethrow l
        | try bt instr* (catch x instr*)+ end
        | try bt instr* (catch x instr*)* catch_all instr* end
        | try bt instr* delegate l
        | try bt instr* unwind instr* end
```

### Modules

Exceptions (definitions)

```
exn ::= export* exception exntype  | export* exception exntype import
```

Modules

```
mod ::= module ... exn*
```

## Validation (Typing)

> To verify that a `try...delegate l` instruction refers to a label introduced by a try block (I'll call this a try-label), introduce a type attribute to labels in the validation context, which type attribute is set to `try` when the label is a try-label. The original notation `label [t*]` will then be a shortcut for `label {result [t*], type ε}`.

Reviewer: There is no need for `delegate` to specifically reference a `try` block. That is, `delegate` does not mean to use a particular handler but rather to ignore the handlers in between. (I suspect `delegate` will likely be implemented either by subtracting code ranges or by popping handlers depending on the backend.) Such a relaxation is also helpful, as one can skip over all handlers, i.e. delegate to the caller, without having to wrap the function body with a `try` and an empty handler.

> Moreover, to verify that the `rethrow l` instruction refers to a label introduced by a catch block (call this a catch-label), we allow this type attribute of labels in the validation context to be set to `catch` when the label is a catch label. This addition is reflected in the execution rules, by the administrative instruction `caught` which introduces a label around the instructions of the catching try-block's catch instructions.

Reviewer 2: So the intention here is, as in the case of `delegate`, `rethrow` can target any label including labels for blocks, loops, and try-delegates, but they are just validation failures, right?

The concept of catch label is a little new. Can we consider it in the same as try label? Then we have this problem:
```wasm
try $label0
  rethrow $label0  ;; cannot be done, because it's not within catch below
catch
end
```

But `delegate` has a similar problem anyway if it takes a try block:
```wasm
try $label0
catch
  try
    ...
  delegate $label0  ;; cannot be done, because $label0's catch is not below but above here
end
```

The problem here is, in reality both instructions target `catch`, not `try`. So what do you think? Should we introduce a concept for catch label for both of them? Then with respect to normal branches, should we say catch label does not exist or the same as its corresponding try label?

> Moreover, to verify that the `rethrow l` instruction refers to a label introduced by a catch block (call this a catch-label), we allow this type attribute of labels in the validation context to be set to `catch` when the label is a catch label. This addition is reflected in the execution rules, by the administrative instruction `caught` which introduces a label around the instructions of the catching try-block's catch instructions.

Reviewer 2: So the intention here is, as in the case of `delegate`, `rethrow` can target any label including labels for blocks, loops, and try-delegates, but they are just validation failures, right?

The concept of catch label is a little new. Can we consider it in the same as try label? Then we have this problem:
```wasm
try $label0
  rethrow $label0  ;; cannot be done, because it's not within catch below
catch
end
```

But `delegate` has another problem anyway if it we say it takes a try block:
```wasm
try $label0
catch
  try
    ...
  delegate $label0  ;; cannot be done, because $label0's catch is not below but above here
end
```

`delegate` can be considered to target a catch label too, but in this case it can be considered to target `try`:
```wasm
try $label1
  try $label0
    try
      ...
    delegate $label0
  delegate $label1
end
```

So my question is
1. For `rethrow`: If we introduce a concept of catch label, should we also mention that that this works in the same way as its corresponding `try` block for other branch-like constructs?
2. For `delegate`: The concept of try block may not work in something like my second example, but as we can see in my third example, it is different from catch label. Should we introduce a new concept?

I might be mistaken on something, so please let me know.

Reviewer 2: Gentle ping :)

Author: Sorry I'm late replying @aheejin!  

> The concept of catch label is a little new. Can we consider it in the same as try label? Then we have this problem:
> 
> ```
> try $label0
>   rethrow $label0  ;; cannot be done, because it's not within catch below
> catch
> end
> ```

I used catch labels just as a helper to make sure that the `rethrow` is inside a catch block, not as something one can use in the text format. In your example, `rethrow` is not inside a catch block, so that example would not validate.

I'm not sure I understand the problem with `delegate`, I'll try to look at this again later today or tomorrow.

Author: @aheejin I see now what you mean, yes, you are right. 

In the second example, the label from the catch block would interfere with `delegate`, when it shouldn't. Also `rethrow` should not interfere with other labels, it should just count the stack of nested catch blocks it is in - to rethrow an exception, an exception must be caught, so we can only `rethrow n` the exception caught by the `n`th surrounding catch block.

I think to solve this, the catch-labels should be removed, and a separate concept of a `caught`-stack should be added to the runtime, described by a parallel to blocks:

```
C^0 ::= val^m [_] instr
C^{n+1} ::= caught{exnaddr val^k} C^n end
```
Also a new stack `caught` to validation contexts, which gets an element "`caught`" prepended whenever we catch an exception by a `catch` block.

About the 3rd example, did you omit a catch in the outer try block on purpose, meaning that `try`s without catches could be allowed, as @tlively suggested in #131 ? In that case, in your 3rd example, if we replace `...` by `throw $exn` then the example would reduce to `throw $exn`. I think it would be a nice simplification to the syntax to allow catchless try blocks. They are equivalent to `try ... catch_all rethrow 0 end` AFAICT, so no new behaviour would be introduced with this simplification.

If we additionally allow `delegate` to target any label, as @RossTate [suggested](https://github.com/WebAssembly/exception-handling/pull/143/files#r522996966), then we can have the same effect also without a try block, as he noted, and we wouldn't have to make any changes to labels. I admit I am not so fond of adding this auxiliary information to labels, I just thought I had to. 

To clarify I'll commit the changes I'm suggesting later today, in three separate patches:

1. add `caught`-stack to the runtime and to validation contexts and remove catch-labels (which actually should have been called "caught" labels but anyway)
2. allow `try`s without `catch`es
3. remove auxiliary `type` of labels and allow `delegate` to reference any label

WDYT?

Author: I just pushed the promised commits. The first one's commit message contains the question "Should the administrative instruction `caught` be removed from the right hand side of rethrow's execution step?", because I ran out of time and I won't be able to think about it until the end of next week.

Author: @aheejin: I just want to clarify that the commits I pushed are just so that it's clear what I meant with my comment above. These commits are independent from each other, so one could be accepted and the next one not without trouble, I think.

Reviewer 2: Sorry for the delayed reply. The catch-less try was a typo, sorry. I didn't mean we have a catchless try. We can maybe consider it, but I personally have a preference against it, as I said in #131. What I tried to convey in that third example was that we can't make `delegate` target outer `catch`, because `delegte` can also target an outer `try-delegate`.

# 3rd proposal formal spec overview

This is an overview of the 3rd proposal's formal spec additions, to aid in discussions concerning the proposed semantics.

## Abstract Syntax

### Types

Exception Types

```
exntype ::= [t*] -> []
```

### Instructions

```
instr ::= ... | throw x | rethrow l
        | try bt instr* (catch x instr*)* (catch_all instr*)? end
        | try bt instr* delegate l
        | try bt instr* unwind instr* end
```

### Modules

Exceptions (definitions)

```
exn ::= export* exception exntype  | export* exception exntype import
```

Modules

```
mod ::= module ... exn*
```

## Validation (Typing)

> To verify that the `rethrow l` instruction refers to a surrounding catch block, we introduce a stack `caught` to validation contexts, which gets an exception index or the keywork `all` prepended whenever we enter instructions inside a `catch exnidx` or `catch_all` block, respectively. This addition is reflected in the execution rules, by the administrative instruction `caught` which models the stack of caught exceptions on the wasm stack.

Reviewer 2: ```suggestion
To verify that the `rethrow l` instruction refers to a surrounding catch block, we introduce a stack `caught` to validation contexts, which gets an exception index or the keyword `all` prepended whenever we enter instructions inside a `catch exnidx` or `catch_all` block, respectively. This addition is reflected in the execution rules, by the administrative instruction `caught` which models the stack of caught exceptions on the wasm stack.
```

### Instructions

```
C_exn(x) = [t*] -> []
--------------------------------
C |- throw x : [t1* t*] -> [t2*]

C_caught(l) =/= ε
-------------------------------
C |- rethrow l : [t1*] -> [t2*]

bt = [t1*] -> [t2*]
C, label [t2*] |- instr* : [t1*] -> [t2*]
(C_exn(x_i) = [t'_i*] -> [])^n
(C, label [t2*], caught x_i |- instr_i* : [t'_i*] -> [t2*])^n
(C, label [t2*], caught all |- instr'* : [] -> [t2*])?
------------------------------------------------------------------
try bt instr* (catch x_i instr_i*)^n (catch_all instr'*)? end : bt

bt = [t1*] -> [t2*]
C, label [t2*] |- instr* : [t1*] -> [t2*]
C_label(l) =/= ε
-----------------------------------------

> try bt instr* delegate l : bt

Reviewer 2: This seems we are using a regular labels for `delegate`. Can we handle this case then?
```wast
try $label0
catch
  try
    ...
  delegate $label0  ;; cannot be done, because $label0's catch is not below but above here
end
```

Is this in line with the @RossTate's [suggestion](https://github.com/WebAssembly/exception-handling/pull/143#discussion_r522996966)? His suggestion looks like we don't need to use labels at all. Am I interpreting it correctly, @RossTate?

Reviewer: That's in line with my suggestion. The number just indicates how many syntactic enclosures to skip out of. A convention in the text format could be that you can label a syntactic enclosure and reference that label in the `delegate` rather than have to compute the number of syntactic enclosures between.

Reviewer 2: @ioannad What do you think? If we are on the same page on this I'll update the explainer to match this.

Author: @aheejin, @RossTate, that sounds good to me!

@aheejin, the example you gave above is now correctly illegal (or nonsensical - not translatable to integer labels), because the additional label attributes are removed (that is, catch-labels are removed, which were wrong, and try blocks are back to using regular labels),

Reviewer 2: Sorry what I meant was not catch label but try label. In my example the label `$label0` is attached to `try`. So it should be a valid try label, but the code itself should be invalid. I'm not sure how you can describe that code using try label, and that's the question I asked.

Did your last commit contain any changes related to this? I don't see any, so...

Author: @aheejin, apologies for the late response. Let me start by saying that I do agree that your example should be impossible. Thank you for clarifying, I see what the problem is, and I also see that we actually do need `caught` to have a label after all, in order to accommodate situations such as:

```
try bt
  throw $e
catch $e
  ....
  br 0 ;; ---
end   ;;    |
;; <------- break to here
```

Tricky situation, I'll have to think about how to express this formally.

Reviewer 2: @ioannad Not sure if I understand. Why is `br 0` a problem here? `br` can use `try` labels just like `block` labels because `try` can also serve as a `block`. Did you try to write `delegate 0` instead in your last example?

Maybe we can just do away with labels for `delegate` definition as @RossTate suggested.

bt = [t1*] -> [t2*]
C, label [t2*] |- instr_1* : [t1*] -> [t2*]

> C, label [t2*], caught all |- instr_2* : [] -> [t2*]
> ----------------------------------------------------
> try bt instr_1* unwind instr_2* : bt

Reviewer: ```suggestion
C, label [t2*] |- instr_2* : [] -> []
----------------------------------------------------
try bt instr_1* unwind instr_2* end : bt
```

This was discussed in #142, but just to reiterate for the sake of having this suggestion be self contained: A purpose of `unwind` is to specify unwinders in a way that works well with non-exceptional non-local control constructs. To that end `rethrow` should not work within `unwind` (which was the case in the earlier formalization), as it may not make sense to `rethrow` other non-local control constructs from within a different dynamic scope. Instead, `unwind` automatically resumes unwinding at the `end`, and so `instr_2*` does not need to output `[t2*]`.

Author: Issue #142 is still open, and (unfortunately) it does not seem to me that there is general agreement on `unwind`. About the return values, in the closed issue #129, it was decided that unwind blocks should indeed return values.

About `rethrow` in `unwind` blocks or not, I can't find that discussion, can you please link to it?

By the way, the earlier formalisation was wrong, as it did not reflect the spec that @aheejin presented and that we voted on.

Author: Thank you for catching the missing `end` typo!

Author: @RossTate, about the `unwind` block's instructions returning values or not: now I see that I was perhaps too quick to close #129, since you don't seem to agree with @aheejin's [proposal there](https://github.com/WebAssembly/exception-handling/issues/129#issuecomment-705800164). Since `unwind` ends with a `rethrow 0`, any return values seem indeed unnecessary. @aheejin what do you think?

Author: @RossTate about `rethrow` inside an `unwind` block's `instr_2*`, I can't see much difference from having a `throw` inside `instr_2*`. Looking around the issues, I only saw arguments saying that a `rethrow` would end the unwinding process, but so would a `br` or `throw` or `return`. I didn't manage to find a discussion wrt `rethrow` of other non-local control constructs. Please help me out find the relevant comments/discussions, I'm a bit lost in the several issues containing mixed information. 

Having said that, please note that the `rethrow 0` resulting from the current `unwind`, is creating a new exception, so _it's throwing within a different dynamic scope,_ IIUC. I do understand that `unwind` should probably be laid out more carefully than what we do here, and this is exactly why I suggested that we postpone it to an extension of the MVP, to avoid making mistakes we can't take back anymore, wrt backwards compatibility. I don't think that people should expect the MVP to be WebAssembly's "final word on exceptions" (or non-local control flow in general - as @phoe suggested), but the "first word on exceptions" (or non-local control flow). :) 

But, as @aheejin  says, I'm also ok to keep `unwind`, if it helps us get the spec settled soon so we can continue our implementations. (Btw, @takikawa and I already started landing first patches for `try-catch` and `throw` in [SpiderMonkey](https://bugzilla.mozilla.org/show_bug.cgi?id=1335652).)

Reviewer 2: @ioannad 
> @RossTate, about the `unwind` block's instructions returning values or not: now I see that I was perhaps too quick to close #129, since you don't seem to agree with @aheejin's [proposal there](https://github.com/WebAssembly/exception-handling/issues/129#issuecomment-705800164). Since `unwind` ends with a `rethrow 0`, any return values seem indeed unnecessary to me as well. @aheejin what do you think? The try block can still have any blocktype, it's just that the unwind block's `instr_2` shouldn't return any values, since they will be discarded by the `rethrow` in the end anyway.

Sorry I don't understand. Throwing values are different from returning values, and they don't have to match. we can branch out of `unwind` block so that we don't end up rethrowing anything, in which case return values make sense, no?

> Having said that, please note that the `rethrow 0` resulting from the current `unwind`, is creating a new exception, so _it's throwing within a different dynamic scope,_ IIUC. 

I thought `rethrow` does not create a new exception, and neither does `end` at the `unwind`..? Why does it create a new exception? If it does, does it preserve metadata like stack traces? I think it should.

Author: @aheejin 

> > @RossTate, about the `unwind` block's instructions returning values or not: now I see that I was perhaps too quick to close #129, since you don't seem to agree with @aheejin's [proposal there](https://github.com/WebAssembly/exception-handling/issues/129#issuecomment-705800164). Since `unwind` ends with a `rethrow 0`, any return values seem indeed unnecessary to me as well. @aheejin what do you think? The try block can still have any blocktype, it's just that the unwind block's `instr_2` shouldn't return any values, since they will be discarded by the `rethrow` in the end anyway.
> 
> Sorry I don't understand. Throwing values are different from returning values, and they don't have to match. we can branch out of `unwind` block so that we don't end up rethrowing anything, in which case return values make sense, no?

I'm not sure I understand. Wouldn't that be branching from "the middle" of the unwind block? I think in that case the part of the unwind block's instructions before the branch should match the branch destination. Then depending on whether we allow return values from unwind blocks or not, we could drop them or adjust them before the end of the unwind block's instructions. Does that make sense?

> 
> > Having said that, please note that the `rethrow 0` resulting from the current `unwind`, is creating a new exception, so _it's throwing within a different dynamic scope,_ IIUC.
> 
> I thought `rethrow` does not create a new exception, and neither does `end` at the `unwind`..? Why does it create a new exception? If it does, does it preserve metadata like stack traces? I think it should.

You're right of course, what I should have said is that `rethrow` is taking an exception possibly originating elsewhere, and throwing it from the context of the `unwind` instruction. So it's not a new exception, but a new throw: rethrow is not propagating the exception, it's throwing it from the location of the rethrow instruction. Does this match your understanding?

Reviewer 2: > @aheejin
> 
> > > @RossTate, about the `unwind` block's instructions returning values or not: now I see that I was perhaps too quick to close #129, since you don't seem to agree with @aheejin's [proposal there](https://github.com/WebAssembly/exception-handling/issues/129#issuecomment-705800164). Since `unwind` ends with a `rethrow 0`, any return values seem indeed unnecessary to me as well. @aheejin what do you think? The try block can still have any blocktype, it's just that the unwind block's `instr_2` shouldn't return any values, since they will be discarded by the `rethrow` in the end anyway.
> > 
> > 
> > Sorry I don't understand. Throwing values are different from returning values, and they don't have to match. we can branch out of `unwind` block so that we don't end up rethrowing anything, in which case return values make sense, no?
> 
> I'm not sure I understand. Wouldn't that be branching from "the middle" of the unwind block? I think in that case the part of the unwind block's instructions before the branch should match the branch destination. Then depending on whether we allow return values from unwind blocks or not, we could drop them or adjust them before the end of the unwind block's instructions. Does that make sense?

I'm not sure why branching from the middle of the unwind block is not allowed? Branching from the middle of a try or a catch block is allowed too, and the return type of those don't have to match the exception type any instruction can throw within the try block. Perhaps I'm not understanding what you are explaining. Can you provide an example?

> 
> > > Having said that, please note that the `rethrow 0` resulting from the current `unwind`, is creating a new exception, so _it's throwing within a different dynamic scope,_ IIUC.
> > 
> > 
> > I thought `rethrow` does not create a new exception, and neither does `end` at the `unwind`..? Why does it create a new exception? If it does, does it preserve metadata like stack traces? I think it should.
> 
> You're right of course, what I should have said is that `rethrow` is taking an exception possibly originating elsewhere, and throwing it from the context of the `unwind` instruction. So it's not a new exception, but a new throw: rethrow is not propagating the exception, it's throwing it from the location of the rethrow instruction. Does this match your understanding?

Yes, I think that makes sense.

Reviewer 3: Sorry to bump this old thread, but I'm not sure a conclusion was reached on the blocktype issue for `unwind`. I'd like to try to clarify with some examples in this thread as all the context is here (but I could open a new issue if that's preferred).

Here are 3 small `unwind` examples. I was wondering which of them should validate:

```
;; Example 1
try (result i32)
  call $f
unwind
  (i32.const 42) ;; matches block type, but value can never be used
end

;; Example 2
try (result i32)
  call $f
unwind
  nop ;; missing i32
end

;; Example 3
try (result)
  call $f
unwind
  (i32.const 42) ;; too many i32s
end
```

My understanding based on the discussion here and the current explainer & spec was that only example 1 should validate, and that examples 2 and 3 are invalid. (in all examples, the stack will be reset by the implicit rethrow of `end` so the values do not matter, but only in 1 is it consistent with the block type)

Does this sound right? (I'm asking for clarification because I think I have implemented this validation slightly differently, so I will go fix those if there's agreement on this)

Reviewer 2: My understanding is also that only the example 1 should validate. Let us know if others have other opinions.

Reviewer 4: I agree that all unwind blocks should be typed `[] -> []`. That's independent of the try's block type, since its a completely separate execution path. So, Example 2 is the only one that ought to type-check.

Reviewer 2: @RossTate What's your take on the typing rule of `unwind` block?

Reviewer 5: Assuming the `[] -> []` type, what is the implication for something like a `br 0` inside the unwind block? Since it exits the block, it needs to leave values on the stack that match the try's signature, so it would be validated using the try block's type rather than the enclosing unwind block's type?

Reviewer 4: @thibaudmichaud, the br is typed according to the label type, which would still be `t2*`.

Reviewer 2: If we can place `br` that returns a value within `unwind`, I'm not sure why the type of `unwind` block has to be `[] -> []`?

Reviewer 4: `br` and the block's `end` are control paths with different targets in the case of unwind, so they have different type constraints (that's also the case with `loop` btw). An unwind block will drop any remaining value on the floor at `end`, so it's pointless to require any values there -- all that would do is force producers to push dummy values for no reason.

Reviewer 6: Sorry I'm late to the party here, but I would expect all three of @takikawa's examples to validate because of implicit unreachable polymorphic behavior due to the implicit `rethrow` at the end of the `unwind` clause, just as they would if the `rethrow` were made explicit as a separate instruction at the end of the block.

Reviewer 4: @tlively, that would be possible as well, but is useless in this case, since unlike branches or throws, an `end` cannot occur in the middle of a larger "expression". Stack polymorphism is useful for examples like
```
(... larger exp (i32.div_u (x) (br_if $bailout (i32.eqz (y))) (y)) ...)
```
or something like that. That cannot be constructed with a delimiter like `end`.

Not saying that we can't do it, but it would be a (minor) relaxation without practical value.

Reviewer 6: I agree there’s no practical value. Not making that relaxation would be fine with me.

        | try bt instr* delegate l
        | try bt instr* unwind instr* end
```

### Modules

Exceptions (definitions)

```
exn ::= export* exception exntype  | export* exception exntype import
```

Modules

```
mod ::= module ... exn*
```

## Validation (Typing)

To verify that a `try...delegate l` instruction refers to a label surrounding the instructions of a try block (call this a try-label), introduce a `kind` attribute to labels in the validation context, which is set to `try` when the label is a try-label.

Similarly, to verify that the `rethrow l` instruction refers to a label surrounding the instructions of a catch block (call this a catch-label), we allow the `kind` attribute of labels in the validation context to be set to `catch` when the label is a catch-label. This addition is reflected in the execution rules, by the administrative instruction `caught` which introduces a label around the catching try-block.

The original notation `label [t*]` is now a shortcut for `label {result [t*], kind ε}`.

### Instructions

```
C_exn(x) = [t*] -> []
--------------------------------
C ⊢ throw x : [t1* t*] -> [t2*]

C_label(l) =/= ε

> C_label(l).kind = catch

Author: done

-------------------------------
C ⊢ rethrow l : [t1*] -> [t2*]

bt = [t1*] -> [t2*]
C, label {result [t2*], kind try} ⊢ instr* : [t1*] -> [t2*]
(C_exn(x_i) = [t'_i*] -> [])^n
(C, label { result [t2*], kind catch } ⊢ instr_i* : [t'_i*] -> [t2*])^n
(C, label { result [t2*], kind catch } ⊢ instr'* : [] -> [t2*])^k

> (k=0 and n>0) or (k=1 and n≥0)

Author: done

------------------------------------------------------------------
C ⊢ try bt instr* (catch x_i instr_i*)^n (catch_all instr'*)^k end : bt

bt = [t1*] -> [t2*]
C, label {result [t2*], kind try} ⊢ instr* : [t1*] -> [t2*]
C_label(l) =/= ε
C_label(l).kind = try
------------------------------------------------------------
C ⊢ try bt instr* delegate l : bt

bt = [t1*] -> [t2*]
C, label {result [t2*], kind try} ⊢ instr_1* : [t1*] -> [t2*]

> C, label [t2*] ⊢ instr_2* : [] -> []

Author: I'm not sure what you mean to "define the new grammar for contexts". Do you mean that in the tba full spec, the text under "Validation" should be expanded but only for labels in throw contexts? 

Btw that text says

> The original notation `label [t*]` is now a shortcut for `label {result [t*], kind ε}`.

But `kind block` sgtm as well. The reason I used the shortcut in the formal rules is to keep them cleaner looking and more resembling the old style. So I only emphasised the different kinds of labels when the kind has implications for other rules.

Do you see a reason to not use the (now) shorthand notation in the formal rules?

> C, label [t2*] ⊢ instr_2* : [] -> []

Reviewer 4: > I'm not sure what you mean to "define the new grammar for contexts".

I just mean that the current spec defines
```
C ::= { ..., label resulttype*, ...}
```
and the new rules seem to assume a change to that, presumably by introducing a new form of `labeltype ::= {result resulttype, kind ??}`. Would help clarity to describe this change here as well.

> Btw that text says
> 
> > The original notation `label [t*]` is now a shortcut for `label {result [t*], kind ε}`.

Ah, sorry, I somehow missed that. Makes sense.

--------------------------------------------------------------
C ⊢ try bt instr_1* unwind instr_2* end : bt
```

## Execution (Reduction)

### Runtime structure

Stores

```
S ::= {..., exn exninst*}
```

Exception Instances

```
exninst ::= {type exntype}
```

Module Instances

```
m ::= {..., (exn a)*}
```

Administrative Instructions

```

> instr ::= ... | throw a | catch_n{a instr*}*{instr*}? instr* end

Reviewer 4: At this level, the ordering constraint on the catch-all handler does not matter, nor that there is only one. So you could simplify the grammar and thus the rules to
```
catch_n{a? instr*}* instr* end
```
where `a? = epsilon` represents catch_all.

Author: done

        | delegate{l} instr* end | caught_m{a val^n} instr* end
```

Throw Contexts

```
T ::= val* [_] instr* | label_n{instr*} T end | caught_m{a val^n} T end | frame_n{F} T end
```

### Instructions

```
F; throw x  ↪  F; throw a  (iff F_exn(x) = a)

caught_m{a val^n} B^{l+1}[rethrow l] end

>   ↪ caught_m{a val^n} B^{l+1}[val^n (throw a)] end

Author: To be honest this has been bothering me lately: as it is now, even in the main spec, any instructions could be between two labels in a `B^l`, or am I misunderstanding something?

Unless you mean that `rethrow`'s immediate `l` should only count the catches, which is [not the case](https://github.com/WebAssembly/exception-handling/issues/146#issuecomment-777349038) anymore (I missed this at first as it's a bit hidden in a different issue).

>   ↪ caught_m{a val^n} B^{l+1}[val^n (throw a)] end

Author: I used `B^{l+1}` because `B^0` does not contain / is not pointing to- a label, while `rethrow 0` should point to a surrounding label.

>   ↪ caught_m{a val^n} B^{l+1}[val^n (throw a)] end

Reviewer 4: > To be honest this has been bothering me lately: as it is now, even in the main spec, any instructions could be between two labels in a `B^l`, or am I misunderstanding something?

Not sure I follow. A branch can skip over any instruction. What would be an example of an odd context?

> I used `B^{l+1}` because `B^0` does not contain / is not pointing to- a label, while `rethrow 0` should point to a surrounding label.

Intuitively, the `B^l` captures the l label binders that you _skip over_ with a branch. So if l=0 then this context should indeed not contain any label. The target label is the one explicit in the rule, surrounding B.

However:

> Also note that, in the execution rules, `caught_m` gets introduced with a label after it (the +1), to compensate for the evaporating try-label when catching the exception by a handler. Does this make sense?

Ah, yes, this explains why we need a +1 here: the catch actually is _outside_ its associated label, so we actually need to skip over that as well to get to the catch.

It's a bit odd, though. I suspect that this use of labels creates an issue with the soundness proof. In particular, what is the administrative typing rule for catch_n? It can't bind the label, since that is bound by label_n. But label_n won't attach the catch kind, so any contained rethrow won't type-check anymore after reducing try to catch, and we lose preservation.

Maybe this can be hacked by changing the reduction rule for try-catch so that it binds the label _outside_ the catch. Then the typing rule for catch_n could _amend_ the label kind of the innermost label in its environment. (And we wouldn't have the +1.)

Something to think about. I'd encourage you to write out the typing rules for the administrative instructions, to check that it all hangs together.

(I guess the more baroque nature of the new design is showing here.)

caught_m{a val^n} val^m end  ↪  val^m
```

A keyword `all` is introduced to simplify the requirements for the `try` execution step below.

```
F; val^n (try bt instr* (catch x_i instr_i*)* (catch_all instr'*)? end)
  ↪  F; catch_m{a_i instr_i*}*{all instr'*}? (label_m{} val^n instr* end) end
  (iff bt = [t1^n] -> [t2^m] and (F_exn(x_i) = a_i)*)

catch_m{a_i instr_i*}*{all instr'*}? val^m end ↪ val^m

> S; F; catch_m{a_i instr_i*}*{all instr'*}? T[val^n (throw a)] end
>   ↪  S; F; caught_m{a val^n} (label_m{} val^n instr_i* end) end
>   (iff S_exn(a) = {type [t^n]->[]} and i is the least such that a_i = a)
> 
> S; F; catch_m{a_i instr_i*}*{all instr*} T[val^n (throw a)] end
>   ↪  S; F; caught_m{a val^n} (label_m instr* end) end
>   (iff S_exn(a) = {type [t^n]->[]} and for every i, a_i =/= a)
> 
> S; F; catch_m{a_i instr_i*}* T[val^n (throw a)] end
>   ↪  S; F; val^n (throw a)
>   (iff for every i, a_i =/= a)

Reviewer 4: You can define these inductively by exhausting the handler list, avoiding the quantifiers in the side conditions. Assuming epsilon for "all":
```suggestion
S; F; catch_m{a1? instr*}{a'? instr'*}* T[val^n (throw a)] end
  ↪  S; F; caught_m{a val^n} (label_m{} val^n instr* end) end
  (iff (a1? = eps \/ a1? = a) /\ S_exn(a) = {type [t^n]->[]})

S; F; catch_m{a1? instr*}{a'? instr'*}* T[val^n (throw a)] end
  ↪  S; F; catch_m{a'? instr'*}* T[val^n (throw a)] end
  (iff a1? =/= eps /\ a1? =/= a)

S; F; catch_m T[val^n (throw a)] end
  ↪  S; F; val^n (throw a)
```

val^n (try bt instr* delegate l) ↪ delegate{l} (label_m{} val^n instr* end) end
  (iff bt = [t^n] -> [t^m])

B^l[ delegate{l} (T[val^n (throw a)]) end ] end

>   ↪ val^n (throw a)

Reviewer 4: I think this is missing a base rule `delegate{l} val^n end ↪ val^n`

Author: added

Reviewer 2: Why does `T` disappear in RHS? If a `label` is in RHS, doesn't that change the depth? Also what does it mean if we have the throwing context is in the form of `caught_m{a val^n} T end` or `frame_n{F} T end`? I think I don't have a good grasp on the role of throwing contexts...

Author: > Why does `T` disappear in RHS? 

Continuing my answer on throw contexts [above](https://github.com/WebAssembly/exception-handling/pull/143/files#r704831771), considering the description:

  _"Throw contexts allow matching the program context around a throw instruction up to the nearest enclosing catch handler, thereby selecting the exception handler responsible for an exception."_

The `T` here matches `throw a` with the surrounding delegate instruction/label, so `delegate {l}` catches and rethrows the thrown exception tag `a`.

> If a `label` is in RHS, doesn't that change the depth?

Inside `T`, "before" `throw a` there are only values or labels, frames, and caught exceptions (no catches or delegates), and "after" `throw a` it's dead code anyway, so I don't think there would be a problem with the depths.

> Also what does it mean if we have the throwing context is in the form of `caught_m{a val^n} T end` or `frame_n{F} T end`? I think I don't have a good grasp on the role of throwing contexts...

I guess these two forms don't match the reduction rule, so the rule can't be applied to them. But `T' = caught_m{a val^n} T end` and `T'' = frame_n{F} T end` are both throw contexts so the rule could apply to these (in case they are immediately surrounded by a `delegate{l} ... end`).
In both cases `delegate` discards the throw context, as well as the surrounding `B^l`, and rethrows the exception.
Am I understanding your question correctly?

Author: Did the comment above somewhat answer your concerns, @aheejin ?

> try bt instr* unwind instr'* end ↪ try bt instr* catch_all instr'* rethrow 0 end

Reviewer 4: :)

# 3rd proposal formal spec overview

This is an overview of the 3rd proposal's formal spec additions, to aid in discussions concerning the proposed semantics.

## Abstract Syntax

### Types

Tag Types

```
tagtype ::= [t*]→[]
```

### Instructions

```
instr ::= ... | throw x | rethrow l
        | try bt instr* (catch x instr*)* (catch_all instr*)? end
        | try bt instr* delegate l
```

### Modules

Tags

```
tag ::= export* tag tagtype  | export* tag tagtype import
```

Modules

```
mod ::= module ... tag*
```

## Validation (Typing)

#### Modification to labels

To verify that a `try...delegate l` instruction refers to a label surrounding the instructions of a try block (call this a try-label), introduce a `kind` attribute to labels in the validation context, which is set to `try` when the label is a try-label.

> Similarly, to verify that the `rethrow l` instruction refers to a label surrounding the instructions of a catch block (call this a catch-label), we allow the `kind` attribute of labels in the validation context to be set to `catch` when the label is a catch-label.

Reviewer 4: To reply for @ioannad: during validation, a try label is used within the try body, a catch label within a catch clause. The former can be delegated to, the latter rethrown from. (Furthermore, if we picked option (3) for delegate, we wouldn't need to distinguish try labels.)

> Similarly, to verify that the `rethrow l` instruction refers to a label surrounding the instructions of a catch block (call this a catch-label), we allow the `kind` attribute of labels in the validation context to be set to `catch` when the label is a catch-label.

Author: Exactly what @rossberg said. I am also fond of option 3 btw, thinking it will simplify the spec. As a followup to this PR I could write the formal overview for option 3. It should be fairly straightforward, but unfortunately I won't have any free time for this during this week.

labelkind ::= try | catch
labeltype ::= {result resulttype, kind labelkind?}
C ::= {..., labels labeltype}
The original notation `labels [t*]` is now an abbreviation for:

```
labels {result [t*], kind ε}
```

### Instructions

```
C.tags[x] = [t*]→[]
-----------------------------
C ⊢ throw x : [t1* t*]→[t2*]

C.labels[l].kind = catch
----------------------------
C ⊢ rethrow l : [t1*]→[t2*]

C.types[bt] = [t1*]→[t2*]
C, labels {result [t2*], kind try} ⊢ instr* : [t1*]→[t2*]
(C.tags[x] = [t*]→[] ∧
 C, labels { result [t2*], kind catch } ⊢ instr* : [t*]→[t2*])*
(C, labels { result [t2*], kind catch } ⊢ instr'* : []→[t2*])?
-----------------------------------------------------------------------------
C ⊢ try bt instr* (catch x instr'*)* (catch_all instr''*)? end : 	[t1*]→[t2*]

C.types[bt] = [t1*]→[t2*]
C, labels {result [t2*], kind try} ⊢ instr* : [t1*]→[t2*]
C.labels[l].kind = try
----------------------------------------------------------
C ⊢ try bt instr* delegate l : [t1*]→[t2*]
```

## Execution (Reduction)

### Runtime structure

Stores

```
S ::= {..., tags taginst*}
```

Tag Instances

```
taginst ::= {type tagtype}
```

Module Instances

```

> m ::= {..., tags a*}

Reviewer 2: What is `a` here? Based on [this section](https://webassembly.github.io/spec/core/exec/runtime.html#module-instances), should we do something like
```
m ::= {..., tags tagtype*}
```
?

Reviewer 4: The grammar should probably use `tagaddr` instead of `a`.

Author: In this case `a`means `tagaddr` (and `m` means `moduleinst`) because this is using the compact format of [this original formal-overview](https://github.com/WebAssembly/exception-handling/issues/87#issuecomment-517216137). The full core formal spec will have `tagaddr` and `moduleinst` there, as well as `blocktype` instead of `bt`, `tagidx` instead of `x`, etc. 

Would you prefer to use the longer format for the overview as well, @aheejin?

Author: @aheejin @rossberg : I'm not sure if you think I should use `tagaddr` instead of `a`, `blocktype` instead of `bt`, `labelidx` instead of `l`, etc, in this document as well, or only in the full spec document.

Reviewer 4: In a _grammar definition_ you have to use the name of the productions, otherwise it would not really be a grammar. In other places, where you just pattern match syntactic phrases whose syntax is already fixed elsewhere, it typically is more convenient to use shorter names. It's a bit like the difference between a data type definition (grammar) where you use the type names and a pattern match (rule) where you have variable names.

Author: Just pushed a commit that addresses this in several places where grammar is defined.

        | delegate{l} instr* end | caught_m{a val^n} instr* end
```

Block contexts and label kinds

So far block contexts are only used in the reduction of `br l` and `return`, and only include labels or values on the stack above the hole. If we want to be able to break jumping over try-catch and try-delegate blocks, we must allow for the new administrative control instructions to appear after labels in block contexts, mirroring the label kinds of labels in validation contexts.

```
B^k ::= val* B'^k instr*
B'^0 ::= [_]
B'^{k+1} ::= label_n{instr*} B^k end | catch_m{a? instr*}* B^{k+1} end | caught_m{a val*} B^{k+1} end | delegate{l} B^{k+1} end
```

Note that `label_n{instr*} label_kind? [_] end? end` could be seen as a simplified control frame.

(Alternatively, we could have the above `label_kind`s be also labels,  remove the additional `label_m` from the execution rules below, and remove the execution rules below where the new administrative instructions only contain `val*`. This would make labels even more similar to control frames.)

Throw Contexts

```
T ::= val* [_] instr* | label_n{instr*} T end | caught_m{a val^n} T end

>    | frame_n{F} T end

Author: IIUC block contexts do not automatically contain other nested constructs (these are the delimited administrative `catch...end`, `delegate...end`, and `caught...end`). I think we have to explicitly add these to the definition of block contexts (lines 120-122 above).  

On the other hand, throw contexts exclude the nested constructs `catch_m { a^? instr* }* [_] end` and `delegate{l} [_] end`, so any throw would have to stop when we encounter the first (innest) `catch_m {...}* [_] end` or `delegate{l} [_] end`.

The prose we used in the 2nd proposal might be clearer:

> Throw contexts allow matching the program context around a throw instruction up to the nearest enclosing `catch` handler, thereby selecting the exception handler responsible for an exception.

>    | frame_n{F} T end

Author: I added a short explanation above the definition of throw contexts. Does that and my comment above clarify their purpose, @aheejin ?

```

### Instructions

```
F; throw x  ↪  F; throw a  (if F.module.tagaddrs[x]=a)

caught_m{a val^n} B^l[rethrow l] end

>   ↪ caught_m{a val^n} B^l[val^n (throw a)] end

Author: I think this doesn't affect the JS API, WDYT @Ms2ger ?

>   ↪ caught_m{a val^n} B^l[val^n (throw a)] end

Author: Do you mean that both `B^l` should be `T`? The intention is that the `l`in `B^l` is matching the one in`rethrow l`, otherwise how will we know which `caught` to target?`Did I miss a spec change on this?

>   ↪ caught_m{a val^n} B^l[val^n (throw a)] end

Author: Can we consider this resolved, or am I missing the error?

caught_m{a val^n} val^m end  ↪  val^m
```

An absent tagaddr in a `catch_m` clause (i.e., `a? = ε`) represents a `catch_all`.

```
F; val^n (try bt instr* (catch x instr'*)* (catch_all instr''*)? end)
  ↪  F; label_m{} (catch_m{a instr'*}*{instr''*}? val^n instr* end) end
  (if bt = [t1^n]→[t2^m] ∧ (F.module.tagaddrs[x]=a)*)

catch_m{a? instr*}* val^m end ↪ val^m

S; F; catch_m{a1? instr*}{a'? instr'*}* T[val^n (throw a)] end
  ↪  S; F; caught_m{a val^n} (val^n)? instr* end
  (if (a1? = ε ∨ a1? = a) ∧ S.tags(a).type = [t^n]→[])

S; F; catch_m{a1? instr*}{a'? instr'*}* T[val^n (throw a)] end
  ↪  S; F; catch_m{a'? instr'*}* T[val^n (throw a)] end
  (if a1? ≠ ε ∧ a1? ≠ a)

S; F; catch_m T[val^n (throw a)] end
  ↪  S; F; val^n (throw a)

val^n (try bt instr* delegate l)

>   ↪ label_m{} (delegate{l+1} val^n instr* end) end

Reviewer 2: Why `l+1`? It would help if we have a brief description on what the administrative instruction `delegate{}` is (not the Wasm instruction `delegate`). The same for the other administrative instructions too

Author: > Why `l+1`?

The `l` becomes `l+1` because we are adding a label around it, so we can break to it from inside `instr*`. 

> It would help if we have a brief description on what the administrative instruction delegate{} is (not the Wasm instruction delegate). The same for the other administrative instructions too

The administrative instructions `catch_n ... end`, `caught_n ... end` and `delegate ... end` are label-like, in that they (plus a surrounding label), delimit an instruction sequence, while also adding information on the special behaviour of these label - label-like combinations.

Off the top of my head perhaps we it's possible to merge each of these with its label, and instead define new labels. So in particular redefine labels to be:

```
label ::= label_n{instr*} |  catch_n{a^? instr*}* |  delegate_n{l} | caught_n{a val*}
```

WDYT?
(cc. @rossberg)

Reviewer 4: Yeah, this is another indication that the hack I suggested for typing catch labels – which is to put the block label outside the catch – is not the greatest idea (see also [this comment]).

I'd much prefer to keep the label statement separate. If you treated all of these as labels, all of them would need to carry a continuation. And likewise future block-constructs we might add. That conflates concerns, increasing complexity.

Unfortunately, there is some conflation due to `rethrow` addressing its source `catch` via a label. (If we hadn't done that, i.e., only count catch clauses, we wouldn't have this problem.) That's why we need the ad-hoc annotation on the label in the context, and that's where the reduction hack comes from.

I'll try to come up with some replacement for this hack – dealing with a typing corner case shouldn't bleed into reduction rules in counter-intuitive ways.

Author: @aheejin, @rossberg, thinking about it, is the +1 really necessary? If we look at `delegate` as:

1. catching an `exn`,
2. "breaking" _inside_ the `l`th surrounding block,
3. and throwing `exn` from there, 

then I think we don't need the `+1`. In fact I think the reduction of administrative delegate is wrong to throw outside of the `l+1`'th surrounding block. WDYT?

Reviewer 4: Hm, you may be right that there is an off-by-1 bug that could work in our favour. Should be easy to check by working through an example. And of course with the interpreter, provided it matches the rules faithfully.

Author: Apologies for the long delay!

Please never mind my previous explanation. I looked at this again and realised that the reduction rule of the administrative `delegate{ l }` should resemble the reduction of `br l`:

```
label_m{} B^l[ delegate{l} T[val^n (throw a)] end ] end
  ↪ val^n (throw a)
```

This way, there is a unique `B^l` that matches the reduction rule.

I modified Example 4 and added Example 5 in `Exceptions-formal-examples.md` which show the reduction steps of the administrative `delegate`s, to check that this works, and modified (or rather simplified) the interpreter accordingly, adding a couple of extra tests there.

I think that now it matches the idea of break-and-throw better. Note that the interpreter part that was removed didn't previously have a corresponding reduction rule in this formal-overview.

# 3rd proposal formal spec overview

This is an overview of the 3rd proposal's formal spec additions, to aid in discussions concerning the proposed semantics.

## Abstract Syntax

### Types

#### Tag Types

```
tagtype ::= [t*]→[]
```

### Instructions

```
instr ::= ... | throw tagidx | rethrow labelidx
        | try blocktype instr* (catch tagidx instr*)* (catch_all instr*)? end
        | try blocktype instr* delegate labelidx
```

### Modules

#### Tags

```
tag ::= export* tag tagtype  | export* tag tagtype import
```

#### Modules

```
mod ::= module ... tag*
```

## Validation (Typing)

#### Modification to labels

To verify that the `rethrow l` instruction refers to a label surrounding the instructions of a catch block (call this a catch-label), we introduce a `kind` attribute to labels in the validation context, which is set to `catch` when the label is a catch-label and empty otherwise.

```
labelkind ::= catch

> labeltype ::= {result resulttype, kind labelkind?}

Author: Nice! I'll use this notation in the next update.

> C ::= {..., labels labeltype}

Author: Will be added in the next update.

```

The original notation `labels [t*]` is now an abbreviation for:

```
labels [t*] ::= labels {result [t*], kind ε}
```

### Instructions

```
C.tags[x] = [t*]→[]
-----------------------------
C ⊢ throw x : [t1* t*]→[t2*]

C.labels[l].kind = catch
----------------------------
C ⊢ rethrow l : [t1*]→[t2*]

C.types[bt] = [t1*]→[t2*]
C, labels [t2*] ⊢ instr* : [t1*]→[t2*]
(C.tags[x] = [t*]→[] ∧
 C, labels { result [t2*], kind catch } ⊢ instr'* : [t*]→[t2*])*
(C, labels { result [t2*], kind catch } ⊢ instr''* : []→[t2*])?
-----------------------------------------------------------------------------
C ⊢ try bt instr* (catch x instr'*)* (catch_all instr''*)? end : [t1*]→[t2*]

C.types[bt] = [t1*]→[t2*]
C, labels [t2*] ⊢ instr* : [t1*]→[t2*]
C.labels[l] = [t*]
-------------------------------------------
C ⊢ try bt instr* delegate l : [t1*]→[t2*]
```

## Execution (Reduction)

### Runtime structure

#### Stores

```
S ::= {..., tags taginst*}
```

#### Tag Instances

```
taginst ::= {type tagtype}
```

#### Module Instances

```
m ::= {..., tags tagaddr*}
```

#### Administrative Instructions

```
instr ::= ... | throw tagaddr | catch_n{ tagaddr? instr* }* instr* end

>         | delegate{ labelidx } instr* end | caught_m{ tagaddr val^n } instr* end

Author: These two instructions are quite different. In short, a `catch_all` reduces to a `catch_m` which only reduces to a `caught_m` if it contains a throw.

In a bit more detail, an instruction `try ... catch_all instr* end` reduces to

```
label_m{} ( catch_m{ ε instr* } ... end ) end
```

Only if the (throw context) `...` contains a `val^n (throw x)` can we reduce to this `caught`:

```
label_m{} ( caught_m{ a_x val^n} instr* end ) end
```

Now `caught` holds the exception that was caught by that `catch_all`, in case a `rethrow` needs to access the caught exception. Note that due to validation `rethrow` can only target `caught` instructions. Does that make sense? Please let me know if this explanation wasn't clear enough.

>         | delegate{ labelidx } instr* end | caught_m{ tagaddr val^n } instr* end

Reviewer 2: I think I understand. I forgot about the possibility of rethrowing. Thanks!

```

#### Block contexts and label kinds

> So far block contexts are only used in the reduction of `br l` and `return`, and only include labels or values on the stack above the hole. If we want to be able to break jumping over try-catch and try-delegate blocks, we must allow for the new administrative control instructions to appear after labels in block contexts.

Author: Apologies, I meant "at the left side of a hole (`[_]`)", I will correct this in the next version.

```
B^0 ::= val* [_] instr*

> B^k ::= catch_m{ tagaddr^? instr* }* B^k end | caught_m{ tagaddr val* } B^k end

Author: Yes, thanks for catching this and all the others, of course!

>       | delegate{ labelidx } B^k end

Reviewer 2: By the way does this support `val*` between `label_n` and `catch_m`/`caught_m`/`delegate`? For example,
```
val* (label_n{instr*} (catch_m{ tagaddr? instr* }* B^k end) end
```
This is supported, but
```
val* (label_n{instr*} val* (catch_m{ tagaddr? instr* }* B^k end) end
```
is this supported? Or should we?

Author: > This is supported, but
> ```
>  val* (label_n{instr*} val* (catch_m{ tagaddr? instr* }* B^k end) end
> ```
> is this supported? Or should we?

The formal-overview does support this, but I can't see any way to reduce to it.
AFAICT I don't think it is a problem to support this case, even if it were possible.

Reviewer 2: My question was, I didn't think the current block context rules can produce 
```
val* (label_n{instr*} val* (catch_m{ tagaddr? instr* }* B^k end) end
```
meaning `val*` between a `label` and `catch`/`caught`/`delegate`. I might have been mistaken. But I think what was suggested in https://github.com/WebAssembly/exception-handling/pull/143#discussion_r759888458 can create this sequence, so I don't think this will be a problem.

> B^{k+1} ::= val* (label_n{instr*} B^k end) instr*

Author: Thanks for the suggestion on the block contexts, I will add it in the next update. 

> Btw, is there a reason that catch and caught have an arity, but delegate does not?

I can't remember any more tbh, although it might be necessary to reduce a `delegate{l} val^n end`. Will add it in the next update as well.

> B^{k+1} ::= val* (label_n{instr*} B^k end) instr*

Reviewer 4: > I can't remember any more tbh, although it might be necessary to reduce a `delegate{l} val^n end`. Will add it in the next update as well.

The `n` there doesn't matter, you could just use `*`. So I don't think it's needed. I am almost sure that `catch` and `caught` do not need it either.

(The reason it's needed for `label` and `frame` is branches and return instructions: the rules for those need to know how many operands to transport along.)

> B^{k+1} ::= val* (label_n{instr*} B^k end) instr*

Author: Thank you for the clarification! That's good, I will remove all of them then. By the way, do you think that using the same names `delegate` and `catch` for the administrative instructions can get confusing? 
What do you think about using `delegation` and `handler' respectively, instead? (Or any other names, these are just examples).

> B^{k+1} ::= val* (label_n{instr*} B^k end) instr*

Reviewer 4: It may be somewhat confusing, but I suspect that inventing different names might just obscure the role of the instructions even more.

FWIW, this question also comes up in the threads proposal, which has more administrative instr's that internalise corresponding source instr's. Would be nice to have a generic naming scheme for this that makes the relation immediately clear. (So far I used `name'` for internal counterparts, but the tick is a bit too low a signal.)

> B^{k+1} ::= val* (label_n{instr*} B^k end) instr*

Reviewer 2: I think the name sharing is fine; coming up with a different set of names just to be different can be awkward and harder to memorize.

```

#### Throw Contexts

Throw contexts don't skip over handlers, they are used to match a thrown exception with the innermost handler.

```
T ::= val* [_] instr* | label_n{instr*} T end | caught_m{ tagaddr val^n } T end
   | frame_n{F} T end
```

Note that because `catch_n` instructions are not included above, there is always a unique maximal throw context.

### Reduction of instructions

Reduction steps for the new instructions or administrative instructions.

```
F; throw x  ↪  F; throw a  (if F.module.tagaddrs[x]=a)

caught_m{a val^n} B^l[rethrow l] end
  ↪ caught_m{a val^n} B^l[val^n (throw a)] end

caught_m{a val^n} val^m end  ↪  val^m
```

An absent tag address in a `handler` clause (i.e., `a? = ε`) represents a `catch_all`.

```
F; val^n (try bt instr* (catch x instr'*)* (catch_all instr''*)? end)
  ↪  F; label_m{} (catch_m{a instr'*}*{ε instr''*}? val^n instr* end) end

>   (if bt = [t1^n]→[t2^m] ∧ (F.module.tagaddrs[x]=a)*)

Reviewer 2: When do we need this `expand_F` and when do we not? It looks this is only necessary in execution semantics, right?

>   (if bt = [t1^n]→[t2^m] ∧ (F.module.tagaddrs[x]=a)*)

Reviewer 4: Yes, in the static semantics, the counterpart is the judgement `C |- bt : ft`, see my [other comment](https://github.com/WebAssembly/exception-handling/pull/143#discussion_r759880458). This all is a bit ugly, but a consequence of our ad-hoc representation of block types.

catch_m{a? instr*}* val^m end ↪ val^m

> S; F; catch_m{a1? instr*}{a'? instr'*}* T[val^n (throw a)] end

Author: In the next patch will remove the `S` and `F` that are not touched in a rule, thanks for pointing this out!

  ↪  S; F; caught_m{a val^n} (val^n)? instr* end
  (if (a1? = ε ∨ a1? = a) ∧ S.tags(a).type = [t^n]→[])

S; F; catch_m{a1? instr*}{a'? instr'*}* T[val^n (throw a)] end
  ↪  S; F; catch_m{a'? instr'*}* T[val^n (throw a)] end
  (if a1? ≠ ε ∧ a1? ≠ a)

S; F; catch_m T[val^n (throw a)] end
  ↪  S; F; val^n (throw a)

val^n (try bt instr* delegate l)
  ↪ label_m{} (delegate{l} val^n instr* end) end
  (if bt = [t^n]→[t^m])

delegate{l} val^n end ↪ val^n

> label_m{} B^l[ delegate{l} T[val^n (throw a)] end ] end

Reviewer 2: I also first wondered about this and wrote a draft of question about that and then I realized this is actually correct after reading examples.. Some comment would be nice.

  ↪ val^n (throw a)
```

Note that the last reduction step above is similar to the reduction of `br`.

> ### Typing rules for administrative instructions

Reviewer 4: @aheejin, these rules are only needed for the soundness proof, so in the spec they appear only in the [Appendix](https://webassembly.github.io/spec/core/appendix/properties.html#administrative-instructions).

```
S.tags[a].type = [t*]→[]
--------------------------------
S;C ⊢ throw a : [t1* t*]→[t2*]

((S.tags[a].type = [t'*]→[])?
 S;C, labels {result [t*], kind catch} ⊢ instr'* : [t'*?]→[t*])*
S;C, labels [t*] ⊢ instr* : []→[t*]
-----------------------------------------------------------------

> S;C, labels [t*] ⊢ catch_n{a? instr'*}* instr* end : []→[t*]

Reviewer 4: @aheejin, these instructions are only produced after any possible parameters have already been popped off the stack (and repushed inside them).

S;C, labels [t*] ⊢ instr* : []→[t*]

> C.labels[l] = [t'*]

Reviewer 4: I believe this should actually be
```suggestion
C.labels[l+1] = [t'*]
```
Otherwise, it would be referring to a different label than the original try-delegate.

Author: @rossberg: I think the +1 is missing because of the behaviour of "breaking inside the label". This way we also cover the case of delegating to the function body. WDYT?

Reviewer 4: Are you sure? If you have
```
try ... delegate $l
```
you'll reduce to
```
label_m{} (delegate{l} val^n instr* end) end
```
When you type this, the outer `label` adds an entry to the labels environment before you get to this rule, unlike with the `try-delegate` before the reduction. So you have to skip over that.

Reviewer 4: The problem I see is that (4) is not the premise you'll need when arriving at the adm.delegate rule. Instead you'll need
```
(4') C'[l'] = [t'*]
```
with `C' = C, labels [t^n]`, because that's the context you now enter the rule with. The only suitable `l'` that allows you to derive this from (4) is `l+1`, I believe.

Reviewer 2: If we change the current rule
```
F; val^n (try bt instr* delegate l)
  ↪ F; label_m{} (delegate{l} val^n instr* end) end
```
to
```
F; val^n (try bt instr* delegate l)
  ↪ F; delegate(l) (label_m{} val^n instr* end) end
```
meaning, if we put the label created by the try-delegate itself within the scope of `adm.delegate`, can this confusion be helped?

Author: Continuing the discussion with a suggestion in #205 .

------------------------------------------------------
S;C, labels [t*] ⊢ delegate{l} instr* end : []→[t*]

S.tags[a].type = [t'*]→[]
(val:t')*

> S;C, labels {result [t*], kind catch} ⊢ instr* : []→[t*]

Author: I think `[t'*]` is only related to the tag type. The empty type there is because `caught` does not consume any values. I think similarly to [this case we discussed](https://github.com/WebAssembly/exception-handling/pull/143/files/9b02debf6a3bb03f64373eb0e56f6f857fc1e68d#r759718940).

----------------------------------------------------------
S;C, labels [t*] ⊢ caught_m{a val^n} instr* end : []→[t*]
```

> 

Reviewer 4: The first one should already be covered by the definition of `T`, which  includes `frame`.

The second is a bit more subtle. We'll need to extend the definition of "terminal", as given in the [Appendix](https://webassembly.github.io/spec/core/appendix/properties.html#theorems) – this has to allow `val* T[throw e] instr*` as a new form of terminal thread configuration, which would be the representation of a computation crashed with an uncaught exception. That does not actually require any additional reduction rule.

(One could try to reduce this to a simpler form like `val* (throw e)`, but that's somewhat tricky, since you'd need a rule that reduces the former to the latter but deterministically only applies at the outermost level. That cannot really be expressed within the current set-up, I believe.)

> 

Reviewer 2: I think addressing this as a followup is fine.

> 

Author: Added a couple of lines in the end, mentioning that we haven't addressed uncaught exceptions yet, as a TODO for a follow up. Resolving this discussion for now.

# 3rd proposal formal spec overview

This is an overview of the 3rd proposal's formal spec additions, to aid in discussions concerning the proposed semantics.

## Abstract Syntax

### Types

Tag Types

```
tagtype ::= [t*]→[]
```

### Instructions

```
instr ::= ... | throw x | rethrow l
        | try bt instr* (catch x instr*)* (catch_all instr*)? end
        | try bt instr* delegate l
```

### Modules

Tags

```

> tag ::= export* tag exntype  | export* tag tagtype import

Reviewer 4: ```suggestion
tag ::= export* tag tagtype  | export* tag tagtype import
```

```

Modules

```
mod ::= module ... tag*
```

## Validation (Typing)

#### Modification to labels

To verify that a `try...delegate l` instruction refers to a label surrounding the instructions of a try block (call this a try-label), introduce a `kind` attribute to labels in the validation context, which is set to `try` when the label is a try-label.

Similarly, to verify that the `rethrow l` instruction refers to a label surrounding the instructions of a catch block (call this a catch-label), we allow the `kind` attribute of labels in the validation context to be set to `catch` when the label is a catch-label.

> 

Reviewer 4: So, in terms of grammar:
```suggestion
labelkind ::= try | catch
labeltype ::= {result resulttype, kind labelkind?}
C ::= {..., labels labeltype}
```
Maybe extend the labelkind with `block` instead of making it an option.

The original notation `labels [t*]` is now an abbreviation for:

```
labels {result [t*], kind ε}
```

### Instructions

```
C.tags[x] = [t*]→[]
-----------------------------
C ⊢ throw x : [t1* t*]→[t2*]

C.labels[l].kind = catch
----------------------------
C ⊢ rethrow l : [t1*]→[t2*]

C.types[bt] = [t1*]→[t2*]
C, labels {result [t2*], kind try} ⊢ instr* : [t1*]→[t2*]
(C.tags[x] = [t*]→[] ∧
 C, labels { result [t2*], kind catch } ⊢ instr* : [t*]→[t2*])*
(C, labels { result [t2*], kind catch } ⊢ instr'* : []→[t2*])?
-----------------------------------------------------------------------------
C ⊢ try bt instr* (catch x instr'*)* (catch_all instr''*)? end : 	[t1*]→[t2*]

C.types[bt] = [t1*]→[t2*]
C, labels {result [t2*], kind try} ⊢ instr* : [t1*]→[t2*]
C.labels[l].kind = try
----------------------------------------------------------
C ⊢ try bt instr* delegate l : [t1*]→[t2*]
```

## Execution (Reduction)

### Runtime structure

Stores

```

> S ::= {..., tags exninst*}

Reviewer 4: ```suggestion
S ::= {..., tags taginst*}
```

```

Tag Instances

```
taginst ::= {type tagtype}
```

Module Instances

```
m ::= {..., tags a*}
```

Administrative Instructions

```
instr ::= ... | throw a | catch_n{a? instr*}* instr* end
        | delegate{l} instr* end | caught_m{a val^n} instr* end
```

Block contexts and label kinds

So far block contexts are only used in the reduction of `br l` and `return`, and only include labels or values on the stack above the hole. If we want to be able to break jumping over try-catch and try-delegate blocks, we must allow for the new administrative control instructions to appear after labels in block contexts, mirroring the label kinds of labels in validation contexts.

```

> label_kind ::= try_label_kind | catch_label_kind
> try_label_kind ::= catch_m{a? instr*}* | delegate{l}
> catch_label_kind ::= caught_m{a val*}
> B^0 ::= val* [_] instr*
> B^{k+1} ::= val* label_n{instr*} label_kind? B^k end? end instr*

Reviewer 4: This seems a bit off, since label_kind and end would be optional independently from each other.
```suggestion
B^k ::= val* B'^k instr*
B'^0 ::= [_]
B'^{k+1} ::= label_n{instr*} B^k end | catch_m{a? instr*}* B^{k+1} end | caught_m{a val*} B^{k+1} end | delegate{l} B^{k+1} end
```

```

Note that `label_n{instr*} label_kind? [_] end? end` could be seen as a simplified control frame.

(Alternatively, we could have the above `label_kind`s be also labels,  remove the additional `label_m` from the execution rules below, and remove the execution rules below where the new administrative instructions only contain `val*`. This would make labels even more similar to control frames.)

Throw Contexts

```
T ::= val* [_] instr* | label_n{instr*} T end | caught_m{a val^n} T end
   | frame_n{F} T end
```

### Instructions

```

> F; throw x  ↪  F; throw a  (if F.module.exnaddrs[x]=a)

Reviewer 4: ```suggestion
F; throw x  ↪  F; throw a  (if F.module.tagaddrs[x]=a)
```

caught_m{a val^n} B^l[rethrow l] end
  ↪ caught_m{a val^n} B^l[val^n (throw a)] end

caught_m{a val^n} val^m end  ↪  val^m
```

> A missing tagaddr in a `catch_m` clause (i.e., `a? = ε`) represents a `catch_all`.

Reviewer 4: ```suggestion
An absent tagaddr in a `catch_m` clause (i.e., `a? = ε`) represents a `catch_all`.
```

```
F; val^n (try bt instr* (catch x instr'*)* (catch_all instr''*)? end)
  ↪  F; label_m{} (catch_m{a instr'*}*{instr''*}? val^n instr* end) end

>   (if bt = [t1^n]→[t2^m] ∧ (val:t1)^n ∧ (F.module.exnaddrs[x]=a)*)

Reviewer 4: ```suggestion
  (if bt = [t1^n]→[t2^m] ∧ (F.module.tagaddrs[x]=a)*)
```
(Note sure what the (val:t1)^n is for. That should always hold due to typing.)

catch_m{a? instr*}* val^m end ↪ val^m

S; F; catch_m{a1? instr*}{a'? instr'*}* T[val^n (throw a)] end
  ↪  S; F; caught_m{a val^n} (val^n)? instr* end
  (if (a1? = ε ∨ a1? = a) ∧ S.tags(a).type = [t^n]→[])

S; F; catch_m{a1? instr*}{a'? instr'*}* T[val^n (throw a)] end
  ↪  S; F; catch_m{a'? instr'*}* T[val^n (throw a)] end
  (if a1? ≠ ε ∧ a1? ≠ a)

S; F; catch_m T[val^n (throw a)] end

>   ↪  S; F; val^n (throw a)

Reviewer 4: Isn't this case already covered by the definition of throw contexts?

val^n (try bt instr* delegate l)
  ↪ label_m{} (delegate{l} val^n instr* end) end
  (if bt = [t^n]→[t^m])

delegate{l} val^n end ↪ val^n

B^l[ delegate{l} T[val^n (throw a)] end ]
  ↪ val^n (throw a)
```

### Typing rules for administrative instructions

```

> S.tags[a].type = [t2*]→[]
> --------------------------------
> S;C ⊢ throw a : [t1* t2*]→[t*]

Reviewer 4: Nit:
```suggestion
S.tags[a].type = [t*]→[]
--------------------------------
S;C ⊢ throw a : [t1* t*]→[t2*]
```

> (S.tags[a].type = [t'*]→[] ∧
>  S;C, labels {result [t*], kind catch} ⊢ instr1* : [t'*]→[t*])*
> (S;C, labels {result [t*], kind catch} ⊢ instr2* : []→[t*])?
> S;C, labels {result [t*], kind try} ⊢ instr3* : []→[t*]
> -----------------------------------------------------------------------
> S;C, labels [t*] ⊢ catch_n{a instr1*}*{instr2*}? instr3* end : []→[t*]

Reviewer 4: This could be simplified somewhat:
```suggestion
((S.tags[a].type = [t'*]→[])?
 S;C, labels {result [t*], kind catch} ⊢ instr'* : [t'*?]→[t*])*
S;C, labels {result [t*], kind try} ⊢ instr* : []→[t*]
-----------------------------------------------------------------------
S;C, labels [t*] ⊢ catch_n{a? instr'*}* instr* end : []→[t*]
```

S;C, labels {result [t*], kind try} ⊢ instr* : []→[t*]
C.labels[l].kind = try
-----------------------------------------------------------------------
S;C, labels [t*] ⊢ delegate{l} instr* end : []→[t*]

S.tags[a].type = [t'*]→[]
(val:t')*
S;C, labels {result [t*], kind catch} ⊢ instr* : []→[t*]
--------------------------------------------------------------------------------

> S;C, labels {result [t*], kind catch} ⊢ caught_m{a val^n} instr* end : []→[t'*]

Reviewer 4: ```suggestion
S;C, labels [t*] ⊢ caught_m{a val^n} instr* end : []→[t'*]
```

```

By adding the attribute `kind` to labels, we are creating situations in the proof of type preservation, where we have a derivation of some `S;C, label [t*] ⊢ instr* : []→[t*]` but we need to have that `S;C, label {result [t*], kind <labelkind>} ⊢ instr* : []→[t*]` for some `<labelkind> ::= try | catch`. To resolve this we add the following typing rule for labels in the context, which ensures our newly introduced `try` and `catch` blocks can contain any instructions a regular block can.

```
S;C, label [t*] ⊢ instr* : []→[t*]
labelkind = try ∨ labelkind = catch
-----------------------------------------------------------
S;C, label {result [t*], kind labelkind} ⊢ instr* : []→[t*]

> ```

Reviewer 4: Hm, can you give an example of why this rule is needed? AFAICS, labelkinds try or catch are strictly more permissive than a regular label, so why would you ever need to forget them?

To verify that the `rethrow l` instruction refers to a label surrounding the instructions of a catch block (call this a catch-label), we introduce a `kind` attribute to labels in the validation context, which is set to `catch` when the label is a catch-label and empty otherwise.

```
labelkind ::= 'catch'
labeltype ::= 'catch'? resulttype
C ::= {..., 'labels' labeltype}
```

The original notation `labels [t*]` is now an abbreviation for:

```
'labels' [t*] ::= 'labels' ε [t*]
```

### Validation Contexts

Validation contexts now hold a list of tag types, one for each tag known to them.
```

> C ::= { ..., 'tags' tagtype*}

Reviewer 4: You also modify the definition of `C` on line 46. Maybe move that here as well?

C ⊢ bt : [t1*]→[t2*]
C, labels [t2*] ⊢ instr* : [t1*]→[t2*]
(C.tags[x] = [t*]→[] ∧

>  C, labels catch [t2*] ⊢ instr'* : [t*]→[t2*])*
> (C, labels catch [t2*] ⊢ instr''* : []→[t2*])?

Reviewer 4: ```suggestion
 C, labels (catch [t2*]) ⊢ instr'* : [t*]→[t2*])*
(C, labels (catch [t2*]) ⊢ instr''* : []→[t2*])?
```

> catch_m{a_i instr_i*}*{all instr'*}? val^m end ↪ val^m

Reviewer 4: ```suggestion
catch_m{a? instr*}* val^m end ↪ val^m
```
